File: backend/biz/handler/expert_alumni_relation.py
"""科技专家校友关系 路由。"""

from fastapi import APIRouter

from application.expert_alumni_relation import ExpertAlumniRelationApplication
from biz.schemas.common import ApiResponse
from biz.schemas.expert_alumni_relation import AlumniRelationQueryRequest

router = APIRouter(prefix="/kg-construction/expert-alumni-relations")
# 兼容前端/文档遗留路径，避免页面或 curl 打到 404
legacy_router = APIRouter(prefix="/kg-service/expert-alumni-relation")
application = ExpertAlumniRelationApplication()


def _describe() -> dict[str, object]:
    return application.describe()


def _query(body: AlumniRelationQueryRequest) -> ApiResponse:
    # service.query 是同步实现（直连 infra graph client）；用 def 让 FastAPI 放进线程池，
    # 每个 worker 可并发处理 ~40 个请求，而不是 async 阻塞事件循环只跑 1 个。
    try:
        result = application.query(
            expert_id=body.expertId,
            target_expert_id=body.targetExpertId,
            school=body.school,
            education_stage=body.educationStage,
            limit=body.limit,
        )
        return ApiResponse(data=result)
    except ValueError as exc:
        return ApiResponse(
            code=400,
            success=False,
            data=None,
            msg=str(exc.args[0]) if exc.args else str(exc),
        )
    except KeyError as exc:
        return ApiResponse(
            code=404,
            success=False,
            data=None,
            msg=str(exc.args[0]) if exc.args else str(exc),
        )


@router.get("")
def describe_expert_alumni_relation() -> dict[str, object]:
    return _describe()


@router.post("/query")
def query_expert_alumni_relation(body: AlumniRelationQueryRequest) -> ApiResponse:
    return _query(body)


@legacy_router.get("")
def legacy_describe_expert_alumni_relation() -> dict[str, object]:
    return _describe()


@legacy_router.post("")
def legacy_query_expert_alumni_relation(body: AlumniRelationQueryRequest) -> ApiResponse:
    return _query(body)

File: frontend/src/api/expertAlumniRelation.ts
/**
 * 科技专家校友关系 API。
 * 后端：/api/v1/kg-construction/expert-alumni-relations
 */

import { http } from './http'

export interface ApiResponse<T> {
  code: number
  success: boolean
  data: T
  msg: string
}

export interface AlumniInteraction {
  coauthorEdge: boolean
  paperCount: number
  patentCount: number
  projectCount: number
  summary: string
}

export interface AlumniItem {
  alumniId: string
  name: string
  sharedInstitutions: string[]
  dimensions: string[]
  educations: unknown[]
  interactions: AlumniInteraction
}

export interface AlumniSummaryRow {
  label: string
  value: string
}

export interface AlumniResultRow extends AlumniSummaryRow {
  tone?: string
}

export interface AlumniRule {
  name: string
  type: string
  target: string
  trigger: string
  logic: string
  output: string
  threshold: string
  audit: string
}

export interface AlumniEntity {
  id: string
  label: string
  entityType: string
  nodeType?: string
  confidence: number
  relations: string
  evidence: string[]
  x?: number
  y?: number
}

export interface AlumniRelation {
  id: string
  from: string
  to: string
  fromName?: string
  toName?: string
  label: string
  category: string
  dimensions?: string[]
  sharedInstitutions?: string[]
  interactions?: AlumniInteraction
}

export interface AlumniProvenanceEvidence {
  title: string
  businessTable: string
  technicalTable: string
  recordId: string
  fieldIdentifier: string
  sourceField?: string
  graphVid?: string
  summary: string
}

export interface AlumniProvenance {
  sourceDatabase: string
  summary?: string
  evidences: AlumniProvenanceEvidence[]
}

export interface AlumniQueryResult {
  expert: { id: string; name: string; educations: unknown[] }
  mode: 'pair' | 'list'
  total: number
  items: AlumniItem[]
  dimensionsCatalog: string[]
  sourceMeta: { space?: string; graph?: string; truncated?: boolean }
  /** 前端结果详情 Tab 对齐字段 */
  summaryRows?: AlumniSummaryRow[]
  resultRows?: AlumniResultRow[]
  evidence?: string[]
  rules?: AlumniRule[]
  entities?: AlumniEntity[]
  relations?: AlumniRelation[]
  graph?: { nodes: AlumniEntity[]; edges: Array<{ id: string; from: string; to: string; label: string; category: string }> }
  provenance?: AlumniProvenance
}

export type AlumniQueryRequest = {
  expertId: string
  targetExpertId?: string
  school?: string
  educationStage?: string
  limit?: number
}

export function describeExpertAlumniRelation() {
  return http.get<Record<string, unknown>>('/v1/kg-construction/expert-alumni-relations')
}

export function queryExpertAlumniRelation(body: AlumniQueryRequest) {
  return http.post<ApiResponse<AlumniQueryResult>>(
    '/v1/kg-construction/expert-alumni-relations/query',
    body,
    { timeout: 60_000 },
  )
}
